File: 115. Distinct Subsequences/dp.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        m, n = len(s), len(t)
        prev = [0] * (n + 1)
        const = [0] * (n + 1)
        const[0] = 1
        prev[0] = 1
        
        for i in range(1, m + 1):
            cur = const[:]
            for j in range(1, n + 1):
                if s[i - 1] == t[j - 1]:
                    cur[j] = prev[j - 1]
                cur[j] += prev[j]
            prev = cur
        return cur[n]

    

File: 115. Distinct Subsequences/dp.rs
impl Solution {
    pub fn num_distinct(s: String, t: String) -> i32 {
        let m = s.len();
        let n = t.len();
        let s = s.as_bytes();
        let t = t.as_bytes();
        let mut prev = vec![0; n + 1];
        prev[0] = 1;
        for i in 1..m + 1 {
            let mut cur = vec![0; n + 1];
            cur[0] = 1;
            for j in 1..n + 1 {
                if s[i - 1] == t[j - 1] {
                    cur[j] = prev[j - 1];
                }
                cur[j] += prev[j];
            }
            prev = cur;
        }
        return prev[n];
    }
}
